velox/docs/develop/testing/fuzzer.rst

```diff
@@ -58,11 +61,60 @@ using OrderBy and StreamingAggregation.
 Fuzzer iterations alternate between generating plans using Values or TableScan
 nodes.
 
+Many functions work well with random input data. However, some functions have
+restrictions on the input values and random data tend to violate these causing
+failures and preventing the fuzzer from exercising the aggregation beyond the
+initial sanity checks.
+
+For example, “min” function has 2 signatures:
+
+.. code-block::
+
+    min(x) → [same as x]
+    Returns the minimum value of all input values.
+
+    min(x, n) → array<[same as x]>
+    Returns n smallest values of all input values of x. n must be a positive integer and not exceed 10,000.
+
+The second signature, let's call it min_n, has 2 arguments. The first argument
+is the value and the second is a constant number of minimum values to return.
+Most of the time, randomly generated value for the second argument doesn’t fail
+into [1, 10’000] range and aggregation fails:
+
+.. code-block::
+
+    VeloxUserError
+    Error Source: USER
+    Error Code: INVALID_ARGUMENT
+    Reason: (3069436511015786487 vs. 10000) second argument of max/min must be less than or equal to 10000
+    Retriable: False
+    Expression: newN <= 10'000
+    Function: checkAndSetN
+    File: /Users/mbasmanova/cpp/velox-1/velox/functions/prestosql/aggregates/MinMaxAggregates.cpp
+    Line: 574
+
+Similarly, approx_distinct function has a signature that allows to specify max
+standard error in the range of [0.0040625, 0.26000]. Random values for 'e' have
+near zero chance to fall into this range.
+
+To enable effective testing of these functions, Aggregation Fuzzer allows
+registering custom input generators for individual functions.
+
 When testing aggregate functions whose results depend on the order of inputs
 (e.g. map_agg, map_union, arbitrary, etc.), the Fuzzer verifies that all plans
 succeed or fail with compatible user exceptions. When plans succeed, the Fuzzer
 verifies that number of result rows is the same across all plans.
 
+Additionally, Fuzzer tests order-sensitive functions using aggregations over
+sorted inputs. When inputs are sorted, the results are deterministic and therefore
+can be verified.
+
+Fuzzer also supports specifying custom result verifiers. For example, array_agg
+results can be verified by first sorting the result arrays. Similarly, map_agg
+results can be partially verified by transforming result maps into sorted arrays
+of map keys. approx_distinct can be verified by comparing the results with
+count(distinct).
```

Comment:
Should we talk about ResultVerifier and that only compare or verify can supported at any one time etc ?

Reply:
Sure. Updated.